node/src/builder.rs

```diff
@@ -0,0 +1,235 @@
+        Some(relay_chain_url) => Ok((
+            Arc::new(RelayChainRPCInterface::new(relay_chain_url).await?) as Arc<_>,
+            None,
+        )),
+        None => build_inprocess_relay_chain(
+            polkadot_config,
+            parachain_config,
+            telemetry_worker_handle,
+            task_manager,
+            hwbench,
+        ),
+    }
+}
+
+/// build parachain nimbus consensus
+pub fn build_nimbus_consensus<RuntimeApi>(
+    id: ParaId,
+    client: Arc<Client<RuntimeApi>>,
+    prometheus_registry: Option<&Registry>,
+    telemetry: Option<TelemetryHandle>,
+    task_manager: &TaskManager,
+    relay_chain_interface: Arc<dyn RelayChainInterface>,
+    transaction_pool: Arc<TransactionPool<RuntimeApi>>,
+    _sync_oracle: Arc<NetworkService<Block, Hash>>,
+    keystore: SyncCryptoStorePtr,
+    force_authoring: bool,
+) -> Result<Box<dyn ParachainConsensus<Block>>, Error>
+where
+    RuntimeApi: ConstructRuntimeApi<Block, Client<RuntimeApi>> + Send + Sync + 'static,
+    RuntimeApi::RuntimeApi: RuntimeApiCommon<StateBackend = StateBackend>
+        + RuntimeApiNimbus
+        + sp_consensus_aura::AuraApi<Block, AuraId>,
+{
+    let spawn_handle = task_manager.spawn_handle();
+    let proposer_factory = sc_basic_authorship::ProposerFactory::with_proof_recording(
+        spawn_handle,
+        client.clone(),
+        transaction_pool,
+        prometheus_registry,
+        telemetry,
+    );
+
+    // NOTE: In nimbus, author_id is unused as it is the RuntimeAPI that identifies the block author
+    let provider = move |_, (relay_parent, validation_data, _author_id)| {
+        let relay_chain_interface = relay_chain_interface.clone();
+        async move {
+            let parachain_inherent =
+                cumulus_primitives_parachain_inherent::ParachainInherentData::create_at(
+                    relay_parent,
+                    &relay_chain_interface,
+                    &validation_data,
+                    id,
+                )
+                .await;
+
+            let time = sp_timestamp::InherentDataProvider::from_system_time();
+
+            let parachain_inherent = parachain_inherent.ok_or_else(|| {
+                Box::<dyn std::error::Error + Send + Sync>::from(
+                    "Failed to create parachain inherent",
+                )
+            })?;
+
+            let nimbus_inherent = nimbus_primitives::InherentDataProvider;
+            Ok((time, parachain_inherent, nimbus_inherent))
+        }
+    };
+
+    Ok(NimbusConsensus::build(BuildNimbusConsensusParams {
+        additional_digests_provider: (),
+        para_id: id,
+        proposer_factory,
+        block_import: client.clone(),
+        parachain_client: client,
+        keystore,
+        skip_prediction: force_authoring,
+        create_inherent_data_providers: provider,
+    }))
+}
+
+/// start standalone dev consensus
+pub fn start_dev_nimbus_instant_seal_consensus<RuntimeApi>(
```

Comment:
shouldnt this be in `service.rs`? It's not a routine that start with `build_`

Reply:
return `Future` here, and used by `service.rs`.